macrosAutoCats/src/main/scala/com/softwaremill/macwire/autocats/MacwireAutoCatsMacros.scala

```diff
@@ -17,118 +19,148 @@ object MacwireCatsEffectMacros {
-    }
+      override def ident: Tree = asResource.ident//????
 
-    def isFactoryMethod(expr: Expr[Any]): Boolean = expr.tree match {
-      // Function with two parameter lists (implicit parameters) (<2.13)
-      case Block(Nil, Function(p, Apply(Apply(f, _), _))) => true
-      case Block(Nil, Function(p, Apply(f, _)))           => true
-      // Function with two parameter lists (implicit parameters) (>=2.13)
-      case Function(p, Apply(Apply(f, _), _)) => true
-      case Function(p, Apply(f, _))           => true
-      // Other types not supported
-      case _ => false
+      override lazy val `type`: Type = typeCheckUtil.typeCheckIfNeeded(value).typeArgs(0)
+      
+      def ttt = `type`
+      lazy val asResource = new Resource(q"cats.effect.kernel.Resource.eval[cats.effect.IO, ${`type`}]($value)") {
+        override lazy val `type`: Type = ttt
+      }
     }
 
-    def isEffect(expr: Expr[Any]): Boolean = {
-      val checkedType = typeCheckUtil.typeCheckIfNeeded(expr.tree)
+    object Effect {
+      def fromTree(tree: Tree): Option[Effect] =
+        if (isEffect(typeCheckUtil.typeCheckIfNeeded(tree))) Some(new Effect(tree))
+        else None
+
+      def underlyingType(tpe: Type): Option[Type] = if(isEffect(tpe)) Some(tpe.typeArgs(0)) else None
 
-      checkedType.typeSymbol.fullName.startsWith("cats.effect.IO") && checkedType.typeArgs.size == 1
+      def isEffect(tpe: Type): Boolean =
+        tpe.typeSymbol.fullName.startsWith("cats.effect.IO") && tpe.typeArgs.size == 1
+        
     }
 
-    val resourcesExprs = dependencies.filter(isResource)
-    val factoryMethodsExprs = dependencies.filter(isFactoryMethod)
-    val effectsExprs = dependencies.filter(isEffect)
-    val instancesExprs = dependencies.diff(resourcesExprs).diff(factoryMethodsExprs).diff(effectsExprs)
+    def providerFromExpr(expr: Expr[Any]): Provider = {
+      val tree = expr.tree
+      (Resource.fromTree(tree) orElse Effect.fromTree(tree) orElse FactoryMethod.fromTree(tree))
+        .getOrElse(new Instance(tree))
+    }
 
-    val resources =
-      (effectsExprs.map(expr => Effect(expr.tree).asResource) ++ resourcesExprs.map(expr => Resource(expr.tree)))
-        .map(r => (r.`type`, r))
-        .toMap
+    def findProviderIn(values: Seq[Value])(tpe: Type): Option[Tree] = values.find(_.`type` <:< tpe).map(_.ident)
 
-    val instances = instancesExprs
-      .map(expr => Instance(expr.tree))
-      .map(i => (i.`type`, i))
-      .toMap
+    val providers = dependencies.map(providerFromExpr)
+    val values: Seq[Value] = {
+      val init: Seq[Value] = providers.flatMap {
+        case e: Effect => Some(e.asResource)
+        case v: Value => Some(v)
+        case _ => None
+      }
 
-    val factoryMethods = factoryMethodsExprs
-      .map(expr => FactoryMethod(expr.tree))
-      .map(i => (i.`type`, i))
-      .toMap
+      providers.foldLeft(init) {
```

Comment:
I'm not quite sure yet what's happening here, but I have a feeling we should do some sorting here. If there are multiple non-value providers (e.g. factories), one factory producing the input for another, the sort order should reflect this.

Another test case to check ;-)

```scala
case class A()
class B
class C

def makeC(b: B): Resource[C] = ...
def makeB(a: A): Resource[B] = ...

autowire[C](makeC _, makeB _)
```

Reply:
My idea was to keep the order of input  dependencies. This kind of sorting may be useful, but when it comes to resource composition I believe that users may want to control the order.

Reply:
Hm I wouldn't say so, it's easy with two resources, but with more resources w/ dependencies it gets complex. You'd have to write this down yourself and perform the sorting. Perfect task for a computer ;)

Reply:
That said, it's a good idea to make the sorting stable, that is maintain the order given by the user if possible, but if dependencies require - reorder as needed

Reply:
Hmm, I see your point, but on the other hand it may be a good idea to make it configurable in future.

Reply:
I'd say that when using a DI library (or framework) one thing I don't want to be concerned with is order of initialization. I'd expect the library to handle this.

And if there's a specific ordering that is required - just make it explicit in code, by adding a dependency (e.g. via a factory method parameter)